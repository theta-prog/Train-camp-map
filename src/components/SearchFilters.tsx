import { useState, useEffect } from 'react'
import { useTranslations } from 'next-intl'

interface SearchFiltersProps {
  onFilterChange: (filters: {
    keyword: string
    maxPrice: number
    facilities: string[]
    activities: string[]
  }) => void
  initialFilters: {
    keyword: string
    maxPrice: number
    facilities: string[]
    activities: string[]
  }
}

const availableFacilities = [
  'toilet',
  'shower',
  'kitchen',
  'rental',
  'shop',
  'parking',
  'wifi'
]

const availableActivities = [
  'hiking',
  'bbq',
  'fishing',
  'canoe',
  'boat',
  'river',
  'stargazing'
]

export default function SearchFilters({ onFilterChange, initialFilters }: SearchFiltersProps) {
  const t = useTranslations()
  const [keyword, setKeyword] = useState(initialFilters.keyword)
  const [maxPrice, setMaxPrice] = useState(initialFilters.maxPrice)
  const [selectedFacilities, setSelectedFacilities] = useState<string[]>(initialFilters.facilities)
  const [selectedActivities, setSelectedActivities] = useState<string[]>(initialFilters.activities)

  useEffect(() => {
    onFilterChange({
      keyword,
      maxPrice,
      facilities: selectedFacilities,
      activities: selectedActivities
    })
  }, [keyword, maxPrice, selectedFacilities, selectedActivities, onFilterChange])

  const handleFacilityChange = (facility: string) => {
    setSelectedFacilities(prev => 
      prev.includes(facility)
        ? prev.filter(f => f !== facility)
        : [...prev, facility]
    )
  }

  const handleActivityChange = (activity: string) => {
    setSelectedActivities(prev => 
      prev.includes(activity)
        ? prev.filter(a => a !== activity)
        : [...prev, activity]
    )
  }

  const clearFilters = () => {
    setKeyword('')
    setMaxPrice(10000)
    setSelectedFacilities([])
    setSelectedActivities([])
    
    // すぐにフィルターを適用
    onFilterChange({
      keyword: '',
      maxPrice: 10000,
      facilities: [],
      activities: []
    })
  }

  return (
    <div className="bg-white rounded-lg shadow-md p-6">
      <div className="flex justify-between items-center mb-4">
        <h2 className="text-lg font-bold text-gray-800">{t('searchFilters.title')}</h2>
        <button
          onClick={clearFilters}
          className="text-sm text-gray-500 hover:text-gray-700"
        >
          {t('searchFilters.clear')}
        </button>
      </div>

      {/* キーワード検索 */}
      <div className="mb-6">
        <label className="block text-sm font-medium text-gray-700 mb-2">
          {t('searchFilters.keyword')}
        </label>
        <input
          type="text"
          value={keyword}
          onChange={(e) => setKeyword(e.target.value)}
          placeholder={t('searchFilters.keywordPlaceholder')}
          className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-green-500"
        />
      </div>

      {/* 価格フィルター */}
      <div className="mb-6">
        <label className="block text-sm font-medium text-gray-700 mb-2">
          {t('searchFilters.maxPrice')}: ¥{maxPrice.toLocaleString()}
        </label>
        <input
          type="range"
          min="1000"
          max="10000"
          step="500"
          value={maxPrice}
          onChange={(e) => setMaxPrice(parseInt(e.target.value))}
          className="w-full h-2 bg-gray-200 rounded-lg appearance-none cursor-pointer"
        />
        <div className="flex justify-between text-xs text-gray-500 mt-1">
          <span>¥1,000</span>
          <span>¥10,000</span>
        </div>
      </div>

      {/* 設備フィルター */}
      <div className="mb-6">
        <label className="block text-sm font-medium text-gray-700 mb-2">
          {t('searchFilters.facilities')}
        </label>
        <div className="space-y-2">
          {availableFacilities.map((facility) => (
            <label key={facility} className="flex items-center">
              <input
                type="checkbox"
                checked={selectedFacilities.includes(facility)}
                onChange={() => handleFacilityChange(facility)}
                className="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
              />
              <span className="ml-2 text-sm text-gray-700">{t(`facilities.${facility}`)}</span>
            </label>
          ))}
        </div>
      </div>

      {/* アクティビティフィルター */}
      <div className="mb-6">
        <label className="block text-sm font-medium text-gray-700 mb-2">
          {t('searchFilters.activities')}
        </label>
        <div className="space-y-2">
          {availableActivities.map((activity) => (
            <label key={activity} className="flex items-center">
              <input
                type="checkbox"
                checked={selectedActivities.includes(activity)}
                onChange={() => handleActivityChange(activity)}
                className="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
              />
              <span className="ml-2 text-sm text-gray-700">{t(`activities.${activity}`)}</span>
            </label>
          ))}
        </div>
      </div>

      <div className="border-t pt-4">
        <p className="text-sm text-gray-600">
          {selectedFacilities.length > 0 || selectedActivities.length > 0 || keyword 
            ? t('searchFilters.matchingCampsites')
            : t('searchFilters.allCampsites')
          }
        </p>
      </div>
    </div>
  )
}
