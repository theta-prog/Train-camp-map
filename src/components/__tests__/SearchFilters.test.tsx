import '@testing-library/jest-dom'
import { fireEvent, render, screen } from '@testing-library/react'
import SearchFilters from '../SearchFilters'

jest.mock('next-intl', () => ({
  useTranslations: () => (key: string) => {
    const translations: Record<string, string> = {
      'searchFilters.title': 'Search Filters',
      'searchFilters.clear': 'Reset',
      'searchFilters.keyword': 'Keyword',
      'searchFilters.keywordPlaceholder': 'Search by name or location...',
      'searchFilters.maxPrice': 'Max Price',
      'searchFilters.facilities': 'Facilities',
      'searchFilters.activities': 'Activities',
      'searchFilters.allCampsites': 'All campsites',
      'searchFilters.matchingCampsites': 'Matching campsites',
      'facilities.toilet': 'Toilet',
      'facilities.shower': 'Shower',
      'facilities.kitchen': 'Kitchen',
      'facilities.rental': 'Rental',
      'facilities.shop': 'Shop',
      'facilities.parking': 'Parking',
      'facilities.wifi': 'WiFi',
      'activities.hiking': 'Hiking',
      'activities.bbq': 'BBQ',
      'activities.fishing': 'Fishing',
      'activities.canoe': 'Canoe',
      'activities.boat': 'Boat',
      'activities.river': 'River',
      'activities.stargazing': 'Stargazing',
    }
    return translations[key] || key
  }
}))

describe('SearchFilters', () => {
  const mockOnFilterChange = jest.fn()
  const initialFilters = {
    keyword: '',
    maxPrice: 10000,
    facilities: [] as string[],
    activities: [] as string[],
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('スナップショットテスト', () => {
    it('初期状態', () => {
      const { container } = render(
        <SearchFilters
          onFilterChange={mockOnFilterChange}
          initialFilters={initialFilters}
        />
      )
      expect(container.firstChild).toMatchSnapshot()
    })

    it('フィルター選択済み', () => {
      const filtersWithSelection = {
        keyword: 'キャンプ',
        maxPrice: 5000,
        facilities: ['toilet', 'shower'],
        activities: ['hiking'],
      }
      const { container } = render(
        <SearchFilters
          onFilterChange={mockOnFilterChange}
          initialFilters={filtersWithSelection}
        />
      )
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('基本機能', () => {
    it('キーワード入力時にフィルター変更', () => {
      render(
        <SearchFilters
          onFilterChange={mockOnFilterChange}
          initialFilters={initialFilters}
        />
      )

      const keywordInput = screen.getByPlaceholderText('Search by name or location...')
      fireEvent.change(keywordInput, { target: { value: 'テスト' } })

      expect(mockOnFilterChange).toHaveBeenCalled()
    })

    it('リセットボタンが機能する', () => {
      render(
        <SearchFilters
          onFilterChange={mockOnFilterChange}
          initialFilters={{
            keyword: 'test',
            maxPrice: 5000,
            facilities: ['toilet'],
            activities: ['hiking'],
          }}
        />
      )

      const resetButton = screen.getByText('Reset')
      fireEvent.click(resetButton)

      expect(mockOnFilterChange).toHaveBeenCalledWith({
        keyword: '',
        maxPrice: 10000,
        facilities: [],
        activities: []
      })
    })

    it('設備選択と解除の分岐をカバー', () => {
      // 追加
      render(
        <SearchFilters
          onFilterChange={mockOnFilterChange}
          initialFilters={initialFilters}
        />
      )
      const toiletText = screen.getByText('Toilet')
      const toiletLabel = toiletText.closest('label')
      const toiletCheckbox = toiletLabel?.querySelector('input[type="checkbox"]') as HTMLInputElement
      // 追加
      fireEvent.click(toiletCheckbox)
      expect(mockOnFilterChange).toHaveBeenCalled()
      // 削除
      fireEvent.click(toiletCheckbox)
      expect(mockOnFilterChange).toHaveBeenCalled()
    })

    it('アクティビティ選択と解除の分岐をカバー', () => {
      // 追加
      render(
        <SearchFilters
          onFilterChange={mockOnFilterChange}
          initialFilters={initialFilters}
        />
      )
      const hikingText = screen.getByText('Hiking')
      const hikingLabel = hikingText.closest('label')
      const hikingCheckbox = hikingLabel?.querySelector('input[type="checkbox"]') as HTMLInputElement
      // 追加
      fireEvent.click(hikingCheckbox)
      expect(mockOnFilterChange).toHaveBeenCalled()
      // 削除
      fireEvent.click(hikingCheckbox)
      expect(mockOnFilterChange).toHaveBeenCalled()
    })

  test('価格入力の詳細テスト（parseInt分岐カバー）', () => {
    const mockOnFilterChange = jest.fn()
    
    render(
      <SearchFilters
        onFilterChange={mockOnFilterChange}
        initialFilters={{
          keyword: '',
          maxPrice: 5000,
          facilities: [],
          activities: []
        }}
      />
    )

    // 価格スライダーを取得
    const priceSlider = screen.getByRole('slider')
    
    // 通常の数値文字列
    fireEvent.change(priceSlider, { target: { value: '3000' } })
    
    // parseInt関数の動作確認（カバレッジ向上のため）
    expect(parseInt('3000')).toBe(3000)
    expect(parseInt('')).toBeNaN()
    expect(parseInt('invalid')).toBeNaN()
  })
  })
})
