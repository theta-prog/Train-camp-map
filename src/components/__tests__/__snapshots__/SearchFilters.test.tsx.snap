// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SearchFilters スナップショットテスト フィルター選択済み 1`] = `
<div
  class="bg-white rounded-lg shadow-md p-6"
>
  <div
    class="flex justify-between items-center mb-4"
  >
    <h2
      class="text-lg font-bold text-gray-800"
    >
      Search Filters
    </h2>
    <button
      class="text-sm text-gray-500 hover:text-gray-700"
    >
      Reset
    </button>
  </div>
  <div
    class="mb-6"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
    >
      Keyword
    </label>
    <input
      class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-green-500"
      placeholder="Search by name or location..."
      type="text"
      value="キャンプ"
    />
  </div>
  <div
    class="mb-6"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
    >
      Max Price
      : ¥
      5,000
    </label>
    <input
      class="w-full h-2 bg-gray-200 rounded-lg appearance-none cursor-pointer"
      max="10000"
      min="1000"
      step="500"
      type="range"
      value="5000"
    />
    <div
      class="flex justify-between text-xs text-gray-500 mt-1"
    >
      <span>
        ¥1,000
      </span>
      <span>
        ¥10,000
      </span>
    </div>
  </div>
  <div
    class="mb-6"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
    >
      Facilities
    </label>
    <div
      class="space-y-2"
    >
      <label
        class="flex items-center"
      >
        <input
          checked=""
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Toilet
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          checked=""
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Shower
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Kitchen
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Rental
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Shop
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Parking
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          WiFi
        </span>
      </label>
    </div>
  </div>
  <div
    class="mb-6"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
    >
      Activities
    </label>
    <div
      class="space-y-2"
    >
      <label
        class="flex items-center"
      >
        <input
          checked=""
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Hiking
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          BBQ
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Fishing
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Canoe
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Boat
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          River
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Stargazing
        </span>
      </label>
    </div>
  </div>
  <div
    class="border-t pt-4"
  >
    <p
      class="text-sm text-gray-600"
    >
      Matching campsites
    </p>
  </div>
</div>
`;

exports[`SearchFilters スナップショットテスト 初期状態 1`] = `
<div
  class="bg-white rounded-lg shadow-md p-6"
>
  <div
    class="flex justify-between items-center mb-4"
  >
    <h2
      class="text-lg font-bold text-gray-800"
    >
      Search Filters
    </h2>
    <button
      class="text-sm text-gray-500 hover:text-gray-700"
    >
      Reset
    </button>
  </div>
  <div
    class="mb-6"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
    >
      Keyword
    </label>
    <input
      class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-green-500"
      placeholder="Search by name or location..."
      type="text"
      value=""
    />
  </div>
  <div
    class="mb-6"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
    >
      Max Price
      : ¥
      10,000
    </label>
    <input
      class="w-full h-2 bg-gray-200 rounded-lg appearance-none cursor-pointer"
      max="10000"
      min="1000"
      step="500"
      type="range"
      value="10000"
    />
    <div
      class="flex justify-between text-xs text-gray-500 mt-1"
    >
      <span>
        ¥1,000
      </span>
      <span>
        ¥10,000
      </span>
    </div>
  </div>
  <div
    class="mb-6"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
    >
      Facilities
    </label>
    <div
      class="space-y-2"
    >
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Toilet
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Shower
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Kitchen
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Rental
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Shop
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Parking
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          WiFi
        </span>
      </label>
    </div>
  </div>
  <div
    class="mb-6"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
    >
      Activities
    </label>
    <div
      class="space-y-2"
    >
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Hiking
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          BBQ
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Fishing
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Canoe
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Boat
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          River
        </span>
      </label>
      <label
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-green-600 focus:ring-green-500 border-gray-300 rounded"
          type="checkbox"
        />
        <span
          class="ml-2 text-sm text-gray-700"
        >
          Stargazing
        </span>
      </label>
    </div>
  </div>
  <div
    class="border-t pt-4"
  >
    <p
      class="text-sm text-gray-600"
    >
      All campsites
    </p>
  </div>
</div>
`;
